rosbag2_py/src/rosbag2_py/_transport.cpp

```diff
@@ -140,53 +142,144 @@ class Player
+    old_sigint_handler_ = std::signal(SIGINT, &rosbag2_py::Player::signal_handler);
+  }
+
+  static void uninstall_signal_handlers()
+  {
+    if (old_sigterm_handler_ != SIG_ERR) {
+      std::signal(SIGTERM, old_sigterm_handler_);
+      old_sigterm_handler_ = SIG_ERR;
+    }
+    if (old_sigint_handler_ != SIG_ERR) {
+      std::signal(SIGINT, old_sigint_handler_);
+      old_sigint_handler_ = SIG_ERR;
+    }
+    deferred_sig_number_ = -1;
+  }
+
+  static void process_deferred_signal()
+  {
+    auto call_signal_handler = [](const SignalHandlerType & signal_handler, int sig_num) {
+        if (signal_handler != SIG_ERR && signal_handler != SIG_IGN && signal_handler != SIG_DFL) {
+          signal_handler(sig_num);
+        }
+      };
+
+    if (deferred_sig_number_ == SIGINT) {
+      call_signal_handler(old_sigint_handler_, deferred_sig_number_);
+    } else if (deferred_sig_number_ == SIGTERM) {
+      call_signal_handler(old_sigterm_handler_, deferred_sig_number_);
+    }
+  }
+
+  void play_impl(
+    const rosbag2_storage::StorageOptions & storage_options,
+    PlayOptions & play_options,
+    bool burst = false,
+    size_t burst_num_messages = 0)
+  {
+    install_signal_handlers();
+    try {
+      auto reader = rosbag2_transport::ReaderWriterFactory::make_reader(storage_options);
+      auto player = std::make_shared<rosbag2_transport::Player>(
+        std::move(reader), storage_options, play_options);
+
+      rclcpp::executors::SingleThreadedExecutor exec;
+      exec.add_node(player);
+      auto spin_thread = std::thread(
+        [&exec]() {
+          exec.spin();
+        });
+      player->play();
+
+      auto wait_for_exit_thread = std::thread(
+        [&]() {
+          std::unique_lock<std::mutex> lock(wait_for_exit_mutex_);
+          wait_for_exit_cv_.wait(lock, [] {return rosbag2_py::Player::exit_.load();});
+          player->stop();
+        });
+      {
+        // Release the GIL for long-running play, so that calling Python code
+        // can use other threads
+        py::gil_scoped_release release;
+        if (burst) {
+          player->burst(burst_num_messages);
+        }
+        player->wait_for_playback_to_finish();
+      }
+
+      rosbag2_py::Player::cancel();  // Need to trigger exit from wait_for_exit_thread
+      if (wait_for_exit_thread.joinable()) {
+        wait_for_exit_thread.join();
+      }
+
+      exec.cancel();
+      if (spin_thread.joinable()) {
+        spin_thread.join();
+      }
+      exec.remove_node(player);
+    } catch (...) {
+      process_deferred_signal();
+      uninstall_signal_handlers();
+      throw;
+    }
+    process_deferred_signal();
+    uninstall_signal_handlers();
   }
+
+  static std::atomic_bool exit_;
+  static std::condition_variable wait_for_exit_cv_;
+  static SignalHandlerType old_sigint_handler_;
+  static SignalHandlerType old_sigterm_handler_;
+  static int deferred_sig_number_;
```

Comment:
I've tried to consider this option but it became very tricky how to maintain the order in which signals were called.  
The tricky part is that it needs to be a static variable. We can't use containers like vectors or queues in static variables since they allocate memory.  
At the end this is all done on `pybind_11` level and unlikely will be important at all if we will call `sigint` than `sigterm`  rather than just one `sigterm`.
I decided to impose a limitation that it will be called only the latest one. Anyway usually both of them will do the same actions just terminate the program.